Feat: Multisig creation with remark

Tbaut opened the pull request:
closes https://github.com/novasamatech/nova-spektr/issues/1756
closes #1807

Tbaut committed:
fix: remove description refactor steps

Tbaut committed:
chore: refactor for useForm

Tbaut committed:
chore: fix flow and confirm/verify

Tbaut committed:
Merge branch 'dev' of github.com:novasamatech/nova-spektr into feat/tbaut-create-multisig

Tbaut committed:
chore: refactor-small-modal

Tbaut committed:
feat: remark signing

Tbaut committed:
feat: most models set

Tbaut committed:
Merge branch 'dev' of github.com:novasamatech/nova-spektr into feat/tbaut-create-multisig

Tbaut committed:
fix: signer selection

Tbaut committed:
feat: signing step

Tbaut committed:
feat: wrapping tx as multi working

Tbaut committed:
chore: revert chainId change

Tbaut committed:
fix: lint

Tbaut committed:
chore: description back

Tbaut committed:
chore: wip

Tbaut committed:
feat: submition ok

Tbaut committed:
feat: some console lgos

Tbaut commented:
There are still quite some things to iron out, such as error management for the form submission, UI glitches, and I'm investigating an issue I found in the multisig indexer, but I'd like to have your opinion on things such as:
- do we need to support proxy signers
- should I organize models differently (right now there's 1 for the flow, 1 for most of the data, 1 for the confirmation). I basically took example on the "add proxy" flow

Tbaut committed:
chore: refactor

Tbaut committed:
fix: lint

Asmadek commented on src/renderer/shared/ui/Modals/BaseModal/BaseModal.tsx:
```diff
@@ -40,10 +40,10 @@ export const BaseModal = ({
-                panelClass,
-              )}
+              className={
+                'transform rounded-lg bg-white text-left align-middle shadow-modal transition-all w-[440px] ' +
```
String concatenation is tricky, use cnTw is preferable, it correctly work when we try to replace some classes from default class.

Tbaut replied:
`cnTw` behaved very weirdly in that case.. and my class (coming from `panelClass`) was actually never appended. Any hint?

Tbaut replied:
I added it back and it worked!

Asmadek commented on src/renderer/widgets/CreateWallet/model/confirm-model.ts:
```diff
@@ -0,0 +1,75 @@
+  { skipVoid: false },
+);
+
+// const $proxiedWallet = combine(
```
Remove commented code or add comment when we should uncomment it

Tbaut replied:
This is a question I asked in the comments, whether the multisig creation should support proxy accounts or not. I may just remove this all, and start simple..

Tbaut replied:
I ended up deleting this :+1:

Asmadek commented on src/renderer/widgets/CreateWallet/model/flow-model.ts:
Now we call main model as widget name, `create-wallet-model` here

Tbaut replied:
This would end up quite verbose though, I copied the behavior of the `AddProxyModal` widget, that also has more or less the same models, and they are named 'form-model', `confirm-model` etc.
Should I rename these `create-wallet-flow-model`, `create-wallet-form-model`...?

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/ConfirmationStep.tsx:
```diff
@@ -0,0 +1,176 @@
+  const selectedSigner = useUnit(flowModel.$selectedSigner);
+  const accountSignatories = useUnit(formModel.$accountSignatories);
+  const wallets = useUnit(walletModel.$wallets);
+  // const proxiedAccounts = useUnit(formModel.$proxiedAccounts);
```
commented code

Tbaut replied:
same.

Tbaut replied:
I ended up deleting this :+1:

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/ConfirmationStep.tsx:
```diff
@@ -0,0 +1,176 @@
+  if (!chain) return null;
+
+  const options = accountSignatories.map(({ accountId, name }) => {
+    //   const isShard = accountUtils.isShardAccount(account);
```
commented code

Tbaut replied:
same, not sure if this is supposed to be supported.

Tbaut replied:
I ended up deleting this :+1:

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/FlexibleMultisigWallet.tsx:
```diff
@@ -59,33 +59,29 @@ export const MultiChainMultisigWallet = ({ isOpen, onClose, onComplete, onBack }
-      setActiveStep((prev) => prev - 1);
-    }
-  };
+  // const goToPrevStep = () => {
```
remove it?

Tbaut replied:
This is all about the Flexible multisig, which is not implemented in this PR. Hence many things commented, before this is done.

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/FlexibleMultisigWallet.tsx:
```diff
@@ -59,33 +59,29 @@ export const MultiChainMultisigWallet = ({ isOpen, onClose, onComplete, onBack }
-  const submitHandler = (args: any) => {
-    toggleResultModal();
-    setName(args.name);
+  // const submitHandler = (args: any) => {
```
remove it?

Tbaut replied:
same here, this is for the flexible multisig

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/MultisigWalletWizard.tsx:
```diff
@@ -0,0 +1,65 @@
+  onComplete: () => void;
+};
+
+// const enum Step {
```
Add comment for commented code or remove it, because now this component looks useless

Tbaut replied:
same here, this will be used shortly, when I tackled the flexible multisig implementation. I discussed this with Yaroslav and I think it's better to just leave it there.

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/NameThresholdStep.tsx:
```diff
@@ -0,0 +1,119 @@
+import { flowModel } from '../../model/flow-model';
+import { Step } from '../../lib/types';
+
+const getThresholdOptions = (optionsAmount: number): DropdownOption<number>[] => {
```
Maybe move this two functions to utils?

Tbaut replied:
Do we use utils for UI components elsewhere? I've taken these 2 functions from the previous component. I can move them to some utils if we usually do this.

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/SelectSignatoriesStep.tsx:
```diff
@@ -0,0 +1,69 @@
+  const hasEnoughSignatories = signatories.length >= 2;
+  const canContinue = hasOwnSignatory && hasEnoughSignatories && !multisigAlreadyExists;
+
+  console.log('<><> canContinue', canContinue);
+  console.log('<><> hasOwnSignatory', hasOwnSignatory);
+
```
do you need this logs?

Tbaut replied:
those are stray, I'll remove.

Tbaut committed:
chore: add tests on form-model

Tbaut committed:
chore: address comments

Tbaut committed:
chore: more tests

Tbaut committed:
chore: test for confirm wallet

Tbaut committed:
chore: flow-model tests

Tbaut committed:
fix: remove most comments

Tbaut commented:
I removed most comments. The multisig creation doesn't take into account proxies for now, and only left the commented logic for the flexible multisig wizard.

Tbaut committed:
fix: revert chains_dev

Tbaut committed:
fix: Update chains_dev.json

Tbaut committed:
fix: Update chains_dev.json

Tbaut committed:
fix: don't show multisigs

Tbaut committed:
Revert "fix: don't show multisigs"

This reverts commit 895234d95c25e94ebc4c4bd78ec98d2686174796.

Tbaut committed:
fix: none step

Tbaut committed:
Revert "Revert "fix: don't show multisigs""

This reverts commit b270c5093c777867ae401853c11235aed1d39a12.

Tbaut committed:
fix: flowfinished to reset

Tbaut committed:
fix: tests

pgolovkin commented:
> There are still quite some things to iron out, such as error management for the form submission, UI glitches, and I'm investigating an issue I found in the multisig indexer, but I'd like to have your opinion on things such as:
> 
> * do we need to support proxy signers
> * should I organize models differently (right now there's 1 for the flow, 1 for most of the data, 1 for the confirmation). I basically took example on the "add proxy" flow

Proxy signatories is not a part of this feature.
The UI should match [mockup](https://www.figma.com/proto/HgFuj8QZd0xqEbkG10DIfs/Flexible-multisig-%7C-Milestone-4?page-id=0:1&node-id=518-88080&viewport=73,491,0.03&t=pRppgwa1U82gQVms-1&scaling=min-zoom&starting-point-node-id=518:88080&show-proto-sidebar=1) because:
1. For multishard wallet, WalletConnect, Nova Wallet the user has to select network before selecting acounts
2. There should be an option to add signatory by address only (without contact book)

Tbaut commented:
May I receive reviews on the content of the PR @tuul-wq please?

Tbaut committed:
fix: fix conflicts

Tbaut committed:
feat: remove Matrix

Tbaut committed:
fix: jest pnpm and lock file

Tbaut committed:
fix: pnpm version

Tbaut committed:
fix: resolve conflicts

Tbaut committed:
chore: merge matrix removal

Tbaut committed:
fix: network and name first

Tbaut committed:
fix: submission

Tbaut commented:
I made the following changes:
- base this PR on top of the matrix removal one https://github.com/novasamatech/nova-spektr/pull/1898 so that we can keep subjects isolated
- I made the change to have first the network selection + name then the signatories selection

Since I did not received any feeback on the effector logic yet, I'd like to start with that. While I'll get the UI to look closer to what the Figma wireframe looks like

Tbaut committed:
feat: wip

Tbaut committed:
feat: style

Tbaut committed:
chore: translation

Tbaut committed:
chore: pull dev

Tbaut committed:
chore: pull feat/tbaut-remove-matrix-bis

Tbaut committed:
fix: lockfile

Tbaut committed:
Merge branch 'feat/tbaut-remove-matrix-bis' of github.com:novasamatech/nova-spektr into feat/tbaut-create-multisig

Tbaut committed:
fix: confirm step

Tbaut committed:
fix: tests

Tbaut committed:
fix: fix tests

Tbaut committed:
fix: fix path

Tbaut committed:
chore: add test.. failing somehow

Tbaut commented on src/renderer/widgets/CreateWallet/model/__tests__/signatory-model.test.ts:
```diff
@@ -0,0 +1,69 @@
+  test('should correctly delete signatories', async () => {
+    const scope = fork();
+
+    expect(scope.getState(signatoryModel.$signatories).size).toEqual(0);
```
Do you guys have any idea how this can actually fail? When run alone, this test succeeds.
When run alone, the test succeeds, when the whole file is ran, it fails, and when observing the `$signatories` store, it's indeed not empty! 
How can I make sure that stores are reset between each test?

@Asmadek @johnthecat @sokolova-an

Asmadek replied:
I think set it directly should work 

```ts
    const scope = fork({
      values: new Map().set(signatoryModel.$signatories, new Map([])),
    });
```

Tbaut replied:
thank you so much, it solved it. I've a hard time understanding why/how it could be non empty though :)

Tbaut committed:
chore: initialize

pgolovkin commented:
@Tbaut I'm getting error when create a multisig wallet.
<img width="1226" alt="image" src="https://github.com/user-attachments/assets/23ae6dac-6d78-43ca-a569-5a007b208226">

Seems that Signer Wallet is undefined
<img width="1224" alt="image" src="https://github.com/user-attachments/assets/b2424c4d-c658-4d7e-935d-a0d7890a2ab1">

But I'm not able to select wallet on the screen where filling signatoies.
<img width="923" alt="image" src="https://github.com/user-attachments/assets/49067592-2464-44d9-bc3d-4cb5bdc72c87">

tuul-wq commented on src/renderer/entities/transaction/lib/callDataDecoder.ts:
```diff
@@ -361,6 +361,9 @@ export const useCallDataDecoder = (): ICallDataDecoder => {
         call: decoded.args[2].toHex(),
       };
     },
+    [TransactionType.REMARK]: (decoded): Record<string, any> => {
```
👍

tuul-wq commented on src/renderer/widgets/CreateWallet/model/__tests__/form-model.test.ts:
Nice tests!

tuul-wq commented on src/renderer/widgets/CreateWallet/model/flow-model.ts:
```diff
@@ -0,0 +1,413 @@
+
+    return {
+      chainId: chain.chainId,
+      address: toAddress(TEST_ACCOUNTS[0], { prefix: 42 }),
```
```suggestion
      address: toAddress(TEST_ACCOUNTS[0], { prefix: SS58_DEFAULT_PREFIX }),
```

tuul-wq commented on src/renderer/widgets/CreateWallet/model/flow-model.ts:
```diff
@@ -0,0 +1,413 @@
+    name,
+    threshold,
+    chainId: chain.chainId,
+    signatories: sortBy(
```
Let's add curly braces, `return` and dedicated variable for readability

tuul-wq commented on src/renderer/widgets/CreateWallet/model/form-model.ts:
```diff
@@ -0,0 +1,113 @@
+sample({
+  clock: $createMultisigForm.submit,
+  source: $createMultisigForm.$isValid,
+  filter: (isValid: boolean) => isValid,
```
I assume `boolean` is not required

Tbaut replied:
I get an error with isValid being implicitely of type any, although `$createMultisigForm.$isValid` is of type `StoreWritable<boolean>`. Maybe something specific with effector-forms?

tuul-wq replied:
Alright, let's leave it as is

tuul-wq commented on src/renderer/widgets/CreateWallet/model/signatory-model.ts:
```diff
@@ -0,0 +1,53 @@
+  clock: signatoriesChanged,
+  source: $signatories,
+  fn: (signatories, { index, name, address }) => {
+    signatories.set(index, { name, address });
```
Modifying original store is not a good idea

Tbaut replied:
I originally thought about having a array, instead of a Map. Now the thing is that arrays have no guarantee of keeping in order (although tbh I've never had issues with it). Maps do. However, this adds a lot of complexity everywhere, since looping through Maps require `Array.from`.

Another complexity that it adds, is that Maps don't trigger a re-rendering of the stores. The only way to trigger is to create a new Map from the original one, modify it, and assign the store to the new map (a new object address).

Do you think I should rather go with arrays, and take the risk of having the array order changed?

Asmadek replied:
Right, create new map -> modify new map -> return new map

Tbaut replied:
ah got it, I made the change.

tuul-wq commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/FlexibleMultisigWallet.tsx:
```diff
@@ -49,7 +46,7 @@ export const MultiChainMultisigWallet = ({ isOpen, onClose, onComplete, onBack }
     if (isOpen && !isModalOpen) {
       toggleIsModalOpen();
-      createMultisigWalletModel.events.reset();
+      // flowModel.events.reset();
```
Serves some purpose?

Tbaut replied:
stray comment, I'll remove. The whole file isn't used at the moment.

tuul-wq commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/NameNetworkSelection.tsx:
```diff
@@ -0,0 +1,89 @@
+          <Button
+            key="create"
+            disabled={!canContinue}
+            onClick={() => flowModel.events.stepChanged(Step.SIGNATORIES_THRESHOLD)}
```
This is triggered in parallel to submit event?

Tbaut replied:
Unless I missed something, there's no submit event triggered here. The flow is cut into several steps, and the submission of the form is only happening in the next step, more specifically, here: https://github.com/novasamatech/nova-spektr/blob/bff125e955be3bada5d0a90250661b40ee9cd2f4/src/renderer/widgets/CreateWallet/ui/MultisigWallet/SelectSignatoriesThreshold.tsx#L103

tuul-wq replied:
Got it, thanks

tuul-wq reviewed (commented):
Few suggestions & questions

tuul-wq commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/components/Signatory.tsx:
```diff
@@ -0,0 +1,111 @@
+
+  const onAddressChange = (newAddress: string) => {
+    try {
+      validateAddress(newAddress);
```
`validateAddress` can throw an error? 
Why not to use `validateAddress` from `utils/address`?

Tbaut replied:
I didn't realize you had one in the utils. Using it now (I'll push after) and it works well, no need to use try/catch any more.

tuul-wq commented on src/renderer/widgets/CreateWallet/lib/create-multisig-utils.ts:
```diff
@@ -0,0 +1,34 @@
+  isNotFirstStep,
+};
+
+function isConfirmStep(step: Step) {
```
@Asmadek I remember we had a talk between me, you and @sokolova-an about some utility function instead of step checker functions. Have anything been done regarding it and @Tbaut should use that function/approach?

Asmadek replied:
[isStep](https://github.com/novasamatech/nova-spektr/blob/a59f223e14813dab5ce2849650b0478c7cb211c0/src/renderer/shared/lib/utils/step.ts#L16) function can be used

Tbaut replied:
Thanks, I re-declared one since my steps are different, but used this approach none the less.

tuul-wq commented on src/renderer/widgets/CreateWallet/model/form-model.ts:
```diff
@@ -0,0 +1,113 @@
+  validateOn: ['submit'],
+});
+
+const formSubmitted = createEvent<void>();
```
Native event to call after form is submitted - `$form.formValidated`
It doesn't cover your case?

Tbaut replied:
oh, I didn't know about this one. I changed it.

tuul-wq commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/MultisigWallet.tsx:
```diff
@@ -1,66 +1,98 @@
 
-    setStep(Step.CREATE_WALLET);
-  };
+  useEffect(() => {
```
Maybe this hook can help - `useModalClose`?

tuul-wq commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/MultisigWallet.tsx:
```diff
@@ -1,66 +1,98 @@
-    setWalletType(undefined);
-    setStep(Step.SELECT_WALLET_TYPE);
-  };
+  if (createMultisigUtils.isSubmitStep(activeStep))
+    return <OperationSubmit isOpen={isModalOpen} onClose={closeMultisigModal} />;
```
```suggestion
  if (createMultisigUtils.isSubmitStep(activeStep)) {
      return <OperationSubmit isOpen={isModalOpen} onClose={closeMultisigModal} />;
  }
```

tuul-wq commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/NameNetworkSelection.tsx:
```diff
@@ -0,0 +1,89 @@
+  } = useForm(formModel.$createMultisigForm);
+
+  const chainOptions = getChainOptions(Object.values(chains));
+  const canContinue = name.isValid;
```
This variable is probably not worth defining for a single use case `disabled={!canContinue}`

tuul-wq reviewed (commented):
Part 2

Asmadek commented on src/renderer/widgets/CreateWallet/lib/create-multisig-utils.ts:
```diff
@@ -0,0 +1,34 @@
+}
+
+function isNotFirstStep(step: Step) {
+  return step > Step.NAME_NETWORK;
```
Better to use `!==` instead `>` for such cases

Tbaut committed:
Update src/renderer/widgets/CreateWallet/model/flow-model.ts

Co-authored-by: Yaroslav Grachev <tuulwq@gmail.com>

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/components/Signatory.tsx:
```diff
@@ -0,0 +1,111 @@
+    }
+  };
+
+  // const options = proxyAccounts.map((proxyAccount) => {
```
add todo when it should be uncommented or remove it

Tbaut replied:
Absolutely, sorry about that, there are still a lot of comments, because for instance I don't know yet how to cater for proxy accounts, or shard accounts etc. I'll remove them.

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/components/SelectSignatories.tsx:
```diff
@@ -1,292 +1,105 @@
-    return !isThisContact && (isSameContactSelected || isSameWalletSelected);
+  }, [contacts.length, chain]);
+
+  //       return !isEthereumAccountId(c.accountId);
```
add todo when it should be uncommented or remove it

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/NameNetworkSelection.tsx:
```diff
@@ -0,0 +1,89 @@
+          </FootnoteText>
+        </div>
+        <div className="flex justify-end items-center mt-auto">
+          <Button
```
looks like `type="submit"` should be here to trigger submit

Tbaut replied:
The submit is triggered in the next step. I have 1 single form, for 2 steps:
1 - Name and network of the multisig
2 - Signatory and threshold selection

only the 2nd is actually submitting.

Asmadek replied:
Should `onSubmit={submitForm}` be in the form component?

Tbaut replied:
I ended up removing any submission at this stage, since they aren't needed.

Tbaut committed:
fix: start addressing some comments

Tbaut committed:
chore: merge dev and fix conflicts

Tbaut committed:
chore: merge upstream

Tbaut committed:
fix: steps

Tbaut committed:
chore: fix lint

Tbaut commented:
@pgolovkin the issue you encountered is probably because the first account is supposed to be one of your own account. This is not something that is enforced yet, and the selection of your wallet doesn't work yet. You can type in some address though, and if you do so, the first account should be yours, then the error disapears, you can type in other signatories etc. Then the signing should work (it did for me, I just tested). The submission to the blockchain failed bc of "wasm trap" but this is not linked to this PR, it's linked to the v12 api unless I missed something.

pgolovkin commented:
> @pgolovkin the issue you encountered is probably because the first account is supposed to be one of your own account. This is not something that is enforced yet, and the selection of your wallet doesn't work yet. You can type in some address though, and if you do so, the first account should be yours, then the error disapears, you can type in other signatories etc. Then the signing should work (it did for me, I just tested). The submission to the blockchain failed bc of "wasm trap" but this is not linked to this PR, it's linked to the v12 api unless I missed something.

thanks for explanation!

pgolovkin commented:
@Tbaut now I'm getting `TooFewSignatories` error. But anyway signing works!
https://westend.subscan.io/extrinsic/21911935-2

Also I'm still getting error when select the account from the Multishard or Polkadot Vault wallet. Seems that there is a problem with multishard and Polkadot Vault wallets (where the wallet may contain may accounts on a single chain).

Tbaut commented:
I'll take care of this now that I know there doesn't seem to be big changes incoming in the architecture. From your Tx, I can see there are indeed 0 signatories, as `other_signatories` is null.

Tbaut committed:
fix: add default signatory

Tbaut committed:
fix: use .formValidated

Tbaut committed:
fix: onSubmit

Tbaut committed:
fix: combobox for own acounts

Tbaut committed:
fix: reset fields

Tbaut committed:
fix: error for empty name

Tbaut committed:
fix: other_signatories empty

Tbaut committed:
fix: test

Tbaut committed:
fix: useModalClose

Tbaut committed:
fix: stray call

Tbaut committed:
fix: manage names and display options of contacts

Tbaut committed:
fix: add signatories as contact

Asmadek commented on src/renderer/widgets/CreateWallet/lib/create-multisig-utils.ts:
```diff
@@ -0,0 +1,20 @@
+ * @param targetStep target step
+ * @return {boolean}
+ */
+function isStep(step: Step, targetStep: Step): boolean {
```
We already have general `isStep` function, maybe use it?

Tbaut replied:
But the steps supported by `isStep` are different. Do you really want me to add my specific steps to the somewhat generic `Step` type? 

```
export const enum Step {
  NONE,
  INIT,
  CONFIRM,
  SIGN,
  SUBMIT,
  BASKET,
}
```

Asmadek replied:
Yes, we decided to do it in such way

Tbaut replied:
I made the `isStep` generic with types, so that anyone can check with their own step. https://github.com/novasamatech/nova-spektr/pull/1800/commits/429570a7754570c9f6250966e9417b3ba8070a6c

This should make things better.

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/FlexibleMultisigWallet.tsx:
```diff
@@ -60,39 +52,26 @@ export const MultiChainMultisigWallet = ({ isOpen, onClose, onComplete, onBack }
-  const submitHandler = (args: { name: string; threshold: number; creatorId: HexString }) => {
-    toggleResultModal();
-    setName(args.name);
+  // const submitHandler = (args: { name: string; threshold: number; creatorId: HexString }) => {
```
commented code

Tbaut replied:
I removed the whole file

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/FlexibleMultisigWallet.tsx:
```diff
@@ -134,20 +106,9 @@ export const MultiChainMultisigWallet = ({ isOpen, onClose, onComplete, onBack }
-              setSignatoryContacts(contacts);
-            }}
-          />
+          {/* {activeStep === Step.CONFIRMATION && <ConfirmationStep />} */}
```
commented code. Check comment line above, is is actual if this code commented?

Tbaut replied:
This whole file isn't used right now. I kept it here because it'll be used soon. But if you prefer I can totally remove it either.

Tbaut replied:
I removed the whole file

Asmadek replied:
Don't forget to remove this file if it's useless, or add comment in code if it can be used in the future

Tbaut replied:
oh thanks, somehow I forgot. 

edit: done in [f0243ca](https://github.com/novasamatech/nova-spektr/pull/1800/commits/f0243ca8b6474d5433a2ae335469b123ce180c7c)

Asmadek commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/MultisigWalletWizard.tsx:
```diff
@@ -0,0 +1,37 @@
+    setTimeout(params?.complete ? onComplete : onClose, DEFAULT_TRANSITION);
+  };
+
+  // const goBack = () => {
```
commented code

Tbaut replied:
I removed the whole file, it wasn't used.

Tbaut committed:
fix: new map then modify it

Tbaut committed:
fix: remove wizzard alltogether

Tbaut committed:
Merge branch 'dev' of github.com:novasamatech/nova-spektr into feat/tbaut-remove-matrix-bis

Tbaut committed:
Merge branch 'feat/tbaut-remove-matrix-bis' of github.com:novasamatech/nova-spektr into feat/tbaut-create-multisig

Tbaut committed:
fix: make isStep generic

Tbaut committed:
fix: lint

Tbaut committed:
chore: merge upstream

Tbaut committed:
chore: fix pnpm.lock

Tbaut committed:
fix: remove chainId change

Tbaut committed:
fix: forgot one

Tbaut committed:
chore: fix conflicts and tests

Tbaut committed:
chore: fix lint

Tbaut committed:
fix: pv not displayed and redirect to multisig

pgolovkin commented:
The list of issues to be fixed:
1. The "multisig" in chains.json must be added only for networks where the multisig indexer is running. So that the list of available networks for the multisig will be limited
<img width="1251" alt="image" src="https://github.com/user-attachments/assets/535e61a9-7e06-42ac-9af8-4c99978e6b97">

2. Validation error should be added only when the user proceeds. Now it's opened when the user gets to the 2nd screen of the Create multisig flow 
<img width="801" alt="image" src="https://github.com/user-attachments/assets/9bc04f2a-fd8d-4d66-bb7b-752eb57b07dc">

3. There is no additional cost (network fee + multisig deposit) that the user has to pay for the `system_remark` call on the 2nd screen of the Create multisig flow. Figma mockup [link](https://www.figma.com/design/HgFuj8QZd0xqEbkG10DIfs/Flexible-multisig-%7C-Milestone-4?node-id=369-58419&t=h20RjewkkjpFoFmn-4)

4. Confirmation screen doesn't match the [mockup]
(https://www.figma.com/design/HgFuj8QZd0xqEbkG10DIfs/Flexible-multisig-%7C-Milestone-4?node-id=528-42184&t=h20RjewkkjpFoFmn-4).

5. Agree, that operation should not be added into the basket.

6. When the user successfully signs `system_remark` operation then the new wallet has to be selected and perations page has to be opened. As I remember it's a common flow for multisig operations, please check it.

Asmadek commented on src/renderer/pages/Basket/lib/prepareTransactions.ts:
```diff
@@ -654,3 +655,28 @@ async function prepareRevokeDelegationTransaction({
+    chain,
+    asset: chain.assets[0],
+    shards: [account],
+    destination: transaction.coreTx.args.dest,
+    remark: '',
```
As I clearly understand from this file `src/renderer/entities/transaction/lib/extrinsicService.ts` remark doesn't have destination param. And why remark param is empty?

pgolovkin replied:
@Asmadek I've created the multisig with this branch and got this transaction https://westend.subscan.io/extrinsic/22813617-2 

I don't know how, but the text for remark was saved on-chain.
<img width="994" alt="image" src="https://github.com/user-attachments/assets/33ce9b93-f7bd-4af2-a1fc-b007f2bc8eb1">

Tbaut replied:
Good point @Asmadek this is not needed since it will never be in the basket. I removed it and excluded to `TransactionData` object type

Tbaut committed:
fix: remove unused file

Tbaut committed:
fix: multisig url and prepare Transaction

Tbaut committed:
fix: remove remark from basket preparation

Tbaut committed:
fix: test

Tbaut committed:
fix: address comment

Tbaut committed:
fix: lint

Tbaut commented:
@pgolovkin I addressed your comments in the last commit.
1. I now take the proxy url when the network has the `multisig` option
2. The alerts are only displayed after the user made a mistake and clicked on next 
3. I now display all additional costs on all screens
4. My screen is now very close to the mock up. The only difference is that it's larger. The reason I think it's better this way, is so that I can re-utilize the nice "Sign with XYZ" button. And this together with the "additional costs" really doesn't look good on small modal, e.g
![image](https://github.com/user-attachments/assets/45cfc703-9bd2-4209-a920-a2ae4b9a3e76)
6. 
> When the user successfully signs system_remark operation then the new wallet has to be selected and perations page has to be opened. As I remember it's a common flow for multisig operations, please check it.

I need help, maybe @Asmadek @sokolova-an or @johnthecat can tell me how to do that.

pgolovkin commented:
> 4\. https://www.figma.com/design/HgFuj8QZd0xqEbkG10DIfs/Flexible-multisig-%7C-Milestone-4?node-id=528-42184&t=h20RjewkkjpFoFmn-4

let's clarify with designer.

pgolovkin commented:
> @pgolovkin I addressed your comments in the last commit.
> 
> 1. I now take the proxy url when the network has the `multisig` option
> 2. The alerts are only displayed after the user made a mistake and clicked on next
> 3. I now display all additional costs on all screens
> 4. My screen is now very close to the mock up. The only difference is that it's larger. The reason I think it's better this way, is so that I can re-utilize the nice "Sign with XYZ" button. And this together with the "additional costs" really doesn't look good on small modal, e.g
>    ![image](https://private-user-images.githubusercontent.com/33178835/372627428-45cfc703-9bd2-4209-a920-a2ae4b9a3e76.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc4NTA0OTMsIm5iZiI6MTcyNzg1MDE5MywicGF0aCI6Ii8zMzE3ODgzNS8zNzI2Mjc0MjgtNDVjZmM3MDMtOWJkMi00MjA5LWE5MjAtYTJhZTRiOWEzZTc2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDAyVDA2MjMxM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTM1MjFjZGFkZTI4NTg3MjQ3MjM1YWYxNDk1YTcwM2U4ZGJlODk3MDQ3ZWU4YWU0M2U4NTZlYzFhMTljMjNiYTAmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.AtvAqU9kKTNj8Znfijai7rfyhwSHq9E8-gOZqFaIMIo)
> 
> > When the user successfully signs system_remark operation then the new wallet has to be selected and perations page has to be opened. As I remember it's a common flow for multisig operations, please check it.
> 
> I need help, maybe @Asmadek @sokolova-an or @johnthecat can tell me how to do that.

1. ok
2. ok
3. ok
4. let's clarify with designer

Tbaut committed:
fix: filter wallets add fees

Tbaut committed:
fix: new design

Tbaut committed:
fix: lint

Tbaut commented:
I've adapted the confirmation step to the [new design](https://www.figma.com/design/HgFuj8QZd0xqEbkG10DIfs/Flexible-multisig-%7C-Milestone-4?node-id=952-47351&t=FJzsx86u7iVtwMR0-4).
The only difference I kept on purpose is the conversion in $ value present on my implementation but not on the design, because I think it brings value to users
![image](https://github.com/user-attachments/assets/60becf1d-e46b-48e4-8777-0f54b86378f1)

pgolovkin commented:
@Tbaut looks great!
 Could you please check how the following scenario works:
1. The user selects more then 1 accounts that the user owns (1 is polkadot vault, 1 is nova wallet)
2. The user proceed 
3. Which account will be selected for signing?

Also please check with Sergy which popup has to be used for showing signatories on the confirmation screen. For example for staking it looks like on the screenshot
<img width="1258" alt="image" src="https://github.com/user-attachments/assets/a48a9e08-7849-4d74-ab9d-6ed5ed246aca">

sokolova-an commented on src/renderer/entities/transaction/ui/MultisigCreationFees.tsx/MultisigCreationFees.tsx:
That looks like a pretty specific file and in entities transaction we keep common things for all transactions. Since it's been used only in CreateWallet maybe move it to CreateWallet/ui/MultisigWallet/components?

Tbaut replied:
Sure, thanks for the suggestion. You mean widget right? I did it in [915c924](https://github.com/novasamatech/nova-spektr/pull/1800/commits/915c924cbf58525f182f7cf161106cdcd0ef2151)

Tbaut committed:
fix: with modal

Tbaut committed:
fix: move MultisigCreationFees to widgets

Tbaut commented:
I implemented the modal for signatories instead of a tooltip: 
![image](https://github.com/user-attachments/assets/d5eb31b8-396b-4e4a-87e7-e5825cce338d)

Tbaut commented:
> The user selects more then 1 accounts that the user owns (1 is polkadot vault, 1 is nova wallet)
    The user proceed
    Which account will be selected for signing?

To answer here, what I answered in tg:
For now, the first account selected has to be an account that is owned, and this is the default signer. Let's discuss the design if you want me to add/change something.

Tbaut committed:
Merge branch 'tbaut-remove-matrix-3rd' of github.com:novasamatech/nova-spektr into feat/tbaut-create-multisig

Tbaut committed:
fix: add signer selection

Tbaut committed:
fix: tests

Tbaut commented:
I added the signer selection, with a "Back" button and without the token info for now. I need some more info to manage the multi-shard accounts. I pinged @johnthecat to have a chat.

Tbaut committed:
fix: support showing pv accounts

Tbaut committed:
fix: title on signer selection

Tbaut committed:
fix: with asset but only the selected one

Tbaut commented on src/renderer/widgets/CreateWallet/ui/MultisigWallet/components/Signer.tsx:
```diff
@@ -0,0 +1,49 @@
+}
+
+export const Signer = ({ accountId, name, onSubmit, chain }: Props) => {
+  //   const balance = useBalance({
```
I have an issue here, where only the balance of the selected account is displayed. If I have 2 possible signers, and I want to display the balance of each of them, the balance is `undefined` for the account that isn't selected.

I tried 2 approaches, one using `useBalance`, the other one fetching the balance directly from `balanceModel`. Both ways failed in the same way.

Can you help @johnthecat @sokolova-an ?

Here is an example, I own both accounts, they both have > 0 WND but only the one that is selected (in blue) has a balance.
![image](https://github.com/user-attachments/assets/3ddae3a6-aff7-47fa-ac53-805fffacd584)

sokolova-an replied:
That's a tricky case, as I remember we don't load balances for all wallets at once, so if you haven't ever opened the wallet - its accounts won't be in the balances storage and you can't get access to the value with balanceModel in this case.
Please check, will it work if before that, you open the second wallet and wait until the balances are loaded?

In any case, I guess, you will have to manually run update or populate balances for the other wallet, because the store data can be outdated anyway. The main logic about balance stores in `balanceSubModel`

Tbaut replied:
Thank you so much for the pointers. Even when I had the other account open before the balance was undefined. I subscribed manually to the balance now, and this fixed it. Thanks again.

Tbaut committed:
fix: sub to wallet balances

Tbaut committed:
fix: wallet icon and name

Tbaut commented:
Now showing the Wallet rather than the accounts @pgolovkin in the signer selection. 
![image](https://github.com/user-attachments/assets/e8991476-7ebb-4107-8915-5a88e6cf87ab)

pgolovkin commented:
@Tbaut looks great! Thanks!

Tbaut committed:
Merge branch 'tbaut-remove-matrix-3rd' of github.com:novasamatech/nova-spektr into feat/tbaut-create-multisig

Tbaut committed:
chore: pnpm lock update

Tbaut committed:
fix: lint - merge gone wrong?
